src/effects/VST3/VST3Effect.cpp

```diff
@@ -0,0 +1,1138 @@
+   return XO("SubCategories: %s").Format( mEffectClassInfo.subCategoriesString() );
+}
+
+EffectType VST3Effect::GetType()
+{
+   using namespace Steinberg::Vst::PlugType;
+   if(mEffectClassInfo.subCategoriesString() == kFxGenerator)
+      return EffectTypeGenerate;
+   const auto& cats = mEffectClassInfo.subCategories();
+   
+   if(std::find(cats.begin(), cats.end(), kFx) != cats.end())
+      return EffectTypeProcess;
+
+   return EffectTypeNone;
+}
+
+EffectFamilySymbol VST3Effect::GetFamily()
+{
+   return XO("VST3");
+}
+
+bool VST3Effect::IsInteractive()
+{
+   return mEditController != nullptr;
+}
+
+bool VST3Effect::IsDefault()
+{
+   return false;
+}
+
+bool VST3Effect::IsLegacy()
+{
+   return false;
+}
+
+bool VST3Effect::SupportsRealtime()
+{
+   return true;
+}
+
+bool VST3Effect::SupportsAutomation()
+{
+   if(mEditController == nullptr)
+      return false;
+
+   using namespace Steinberg;
+   
+   for(int i = 0, count = mEditController->getParameterCount(); i < count; ++i)
+   {
+      Vst::ParameterInfo parameterInfo;
+      if(mEditController->getParameterInfo(i, parameterInfo) == kResultOk)
+      {
+         if(parameterInfo.flags & Vst::ParameterInfo::kCanAutomate)
+            return true;
+      }
+   }
+
+   return false;
+}
+
+bool VST3Effect::GetAutomationParameters(CommandParameters& parms)
+{
+   if(mEditController == nullptr)
+      return false;
+
+   using namespace Steinberg;
+   
+   for(int i = 0, count = mEditController->getParameterCount(); i < count; ++i)
+   {
+      Vst::ParameterInfo parameterInfo;
+      if(mEditController->getParameterInfo(i, parameterInfo) == kResultOk)
+      {
+         parms.Write(
+            std::to_string(parameterInfo.id),
+            mEditController->getParamNormalized(parameterInfo.id)
+         );
+      }
+   }
+
+   return true;
+}
+
+bool VST3Effect::SetAutomationParameters(CommandParameters& parms)
+{
+   using namespace Steinberg;
+
+   if(mComponentHandler == nullptr)
+      return false;
+
+   for(int i = 0, count = mEditController->getParameterCount(); i < count; ++i)
+   {
+      Vst::ParameterInfo parameterInfo;
+      if(mEditController->getParameterInfo(i, parameterInfo) != kResultOk)
+         continue;
+
+      Vst::ParamValue value;
+      if(parms.Read(std::to_string(parameterInfo.id), &value))
+      {
+         if(mComponentHandler->beginEdit(parameterInfo.id) == kResultOk)
+         {
+            auto cleanup = finally([&]{
+               mComponentHandler->endEdit(parameterInfo.id);
+            });
+            mComponentHandler->performEdit(parameterInfo.id, value);
+         }
+         mEditController->setParamNormalized(parameterInfo.id, value);
+      }
+   }
+
+   return true;
+}
+
+bool VST3Effect::LoadUserPreset(const RegistryPath& name)
+{
+   using namespace Steinberg;
+
+   if(!mEditController)
+      return false;
+
+   if(!PluginSettings::HasConfigValue(*this, PluginSettings::Private, name, wxT("ProcessorState")))
```

Comment:
I would prefer the string literals "ProcessorState" and "ControllerState" appearing in only one place each and given a name

Reply:
This would be inconsistent with other use cases of PluginSettings then